File: ha-git-ops/src/reconciler.go
package main

// Reconciler state machine per managed file:
//
//	in-sync    live == HEAD
//	drift      live != HEAD and HEAD did not move (click-ops change)
//	conflict   HEAD moved AND live != previously-applied version
//	           (both sides changed; never auto-resolved)
//
// Apply only writes files whose live content still matches the previously
// applied git version — local edits are never stomped. On first run
// nothing is written: remote HEAD becomes the baseline and every
// difference shows up as drift for the human to promote or revert.

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	secretsSops  = "secrets.sops.yaml"
	secretsPlain = "secrets.yaml"
)

// Files whose change requires a core restart rather than a reload.
var restartFiles = map[string]bool{
	"configuration.yaml": true,
	secretsPlain:         true,
}

var reloadServices = map[string][2]string{
	"automations.yaml": {"automation", "reload"},
	"scenes.yaml":      {"scene", "reload"},
	"scripts.yaml":     {"script", "reload"},
}

type Status struct {
	Drift           map[string]string
	Conflicts       map[string]string
	Error           string
	RestartRequired bool
	LastSync        string
}

type persistedState struct {
	AppliedSHA string `json:"applied_sha"`
}

type Reconciler struct {
	opts       Options
	ha         *HA
	repoDir    string
	configDir  string
	stateFile  string
	ageKeyFile string
	sshKeyFile string
	backupsDir string

	AgePub string
	SSHPub string

	mu     sync.Mutex
	state  persistedState
	status Status
}

func NewReconciler(opts Options, ha *HA) *Reconciler {
	r := &Reconciler{
		opts:       opts,
		ha:         ha,
		repoDir:    "/data/repo",
		configDir:  "/homeassistant",
		stateFile:  "/data/state.json",
		ageKeyFile: "/data/age.key",
		sshKeyFile: "/data/ssh/id_ed25519",
		backupsDir: "/data/backups",
		status: Status{
			Drift:     map[string]string{},
			Conflicts: map[string]string{},
		},
	}
	if raw, err := os.ReadFile(r.stateFile); err == nil {
		_ = json.Unmarshal(raw, &r.state)
	}
	return r
}

// Snapshot returns a copy of the current status and applied SHA for the UI.
func (r *Reconciler) Snapshot() (Status, string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	st := r.status
	st.Drift = copyMap(r.status.Drift)
	st.Conflicts = copyMap(r.status.Conflicts)
	return st, r.state.AppliedSHA
}

func (r *Reconciler) SetError(msg string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.status.Error = msg
}

func copyMap(m map[string]string) map[string]string {
	out := make(map[string]string, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

func (r *Reconciler) saveState() error {
	raw, _ := json.Marshal(r.state)
	return os.WriteFile(r.stateFile, raw, 0o644)
}

// ---------- setup ----------

func (r *Reconciler) EnsureKeys() error {
	if _, err := os.Stat(r.ageKeyFile); err != nil {
		if out, err := exec.Command("age-keygen", "-o", r.ageKeyFile).CombinedOutput(); err != nil {
			return fmt.Errorf("age-keygen: %v: %s", err, out)
		}
		_ = os.Chmod(r.ageKeyFile, 0o600)
	}
	pub, err := exec.Command("age-keygen", "-y", r.ageKeyFile).Output()
	if err != nil {
		return fmt.Errorf("age-keygen -y: %v", err)
	}
	r.AgePub = strings.TrimSpace(string(pub))

	if _, err := os.Stat(r.sshKeyFile); err != nil {
		if err := os.MkdirAll(filepath.Dir(r.sshKeyFile), 0o700); err != nil {
			return err
		}
		if out, err := exec.Command("ssh-keygen", "-t", "ed25519", "-N", "",
			"-C", "ha-git-ops", "-f", r.sshKeyFile).CombinedOutput(); err != nil {
			return fmt.Errorf("ssh-keygen: %v: %s", err, out)
		}
	}
	sshPub, err := os.ReadFile(r.sshKeyFile + ".pub")
	if err != nil {
		return err
	}
	r.SSHPub = strings.TrimSpace(string(sshPub))

	log.Printf("age public key (add to .sops.yaml recipients, then `sops updatekeys`): %s", r.AgePub)
	log.Printf("ssh deploy key (add to the git host with write access): %s", r.SSHPub)
	return nil
}

// ---------- git plumbing ----------

func (r *Reconciler) gitEnv() []string {
	return append(os.Environ(),
		"GIT_SSH_COMMAND=ssh -i "+r.sshKeyFile+" -o StrictHostKeyChecking=accept-new",
		"HOME=/data",
	)
}

func (r *Reconciler) git(args ...string) (string, error) {
	cmd := exec.Command("git", append([]string{"-C", r.repoDir}, args...)...)
	cmd.Env = r.gitEnv()
	var out, errb bytes.Buffer
	cmd.Stdout, cmd.Stderr = &out, &errb
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("git %s: %s", strings.Join(args, " "),
			strings.TrimSpace(errb.String()))
	}
	return out.String(), nil
}

func (r *Reconciler) ensureRepo() error {
	if _, err := os.Stat(filepath.Join(r.repoDir, ".git")); err == nil {
		return nil
	}
	if r.opts.RepoURL == "" {
		return fmt.Errorf("repo_url is not configured")
	}
	cmd := exec.Command("git", "clone", "--filter=blob:none", "--no-checkout",
		"--branch", r.opts.Branch, r.opts.RepoURL, r.repoDir)
	cmd.Env = r.gitEnv()
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("git clone: %v: %s", err, out)
	}
	for _, args := range [][]string{
		{"sparse-checkout", "set", r.opts.Subfolder},
		{"checkout", r.opts.Branch},
		{"config", "user.name", r.opts.CommitName},
		{"config", "user.email", r.opts.CommitEmail},
	} {
		if _, err := r.git(args...); err != nil {
			return err
		}
	}
	return nil
}

// atRef returns file content at a git ref (nil if absent). The secrets
// file is decrypted, so callers always see the live-file shape.
func (r *Reconciler) atRef(ref, rel string) []byte {
	src := rel
	if rel == secretsPlain {
		src = secretsSops
	}
	cmd := exec.Command("git", "-C", r.repoDir, "show",
		ref+":"+r.opts.Subfolder+"/"+src)
	cmd.Env = r.gitEnv()
	out, err := cmd.Output()
	if err != nil {
		return nil
	}
	if rel == secretsPlain {
		dec, err := r.decrypt(out)
		if err != nil {
			log.Printf("sops decrypt failed: %v", err)
			return nil
		}
		return dec
	}
	return out
}

func (r *Reconciler) decrypt(blob []byte) ([]byte, error) {
	cmd := exec.Command("sops", "--decrypt",
		"--input-type", "yaml", "--output-type", "yaml", "/dev/stdin")
	cmd.Env = append(os.Environ(),
		"SOPS_AGE_KEY_FILE="+r.ageKeyFile, "HOME=/data")
	cmd.Stdin = bytes.NewReader(blob)
	var out, errb bytes.Buffer
	cmd.Stdout, cmd.Stderr = &out, &errb
	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("%s", strings.TrimSpace(errb.String()))
	}
	return out.Bytes(), nil
}

// desiredRels lists managed rel-paths at a ref, in live-file naming.
func (r *Reconciler) desiredRels(ref string) ([]string, error) {
	out, err := r.git("ls-tree", "-r", "--name-only", ref, "--", r.opts.Subfolder)
	if err != nil {
		return nil, err
	}
	var rels []string
	prefix := r.opts.Subfolder + "/"
	for _, p := range strings.Split(strings.TrimSpace(out), "\n") {
		if p == "" {
			continue
		}
		rel := strings.TrimPrefix(p, prefix)
		if rel == secretsSops {
			rel = secretsPlain
		}
		rels = append(rels, rel)
	}
	return rels, nil
}

func (r *Reconciler) live(rel string) []byte {
	b, err := os.ReadFile(filepath.Join(r.configDir, rel))
	if err != nil {
		return nil
	}
	return b
}

// ---------- reconcile ----------

func (r *Reconciler) Tick() error {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.tickLocked()
}

func (r *Reconciler) tickLocked() error {
	if err := r.ensureRepo(); err != nil {
		return err
	}
	if _, err := r.git("fetch", "origin", r.opts.Branch); err != nil {
		return err
	}
	remote, err := r.git("rev-parse", "origin/"+r.opts.Branch)
	if err != nil {
		return err
	}
	remote = strings.TrimSpace(remote)

	switch {
	case r.state.AppliedSHA == "":
		// First run: baseline only, write nothing. Differences appear as
		// drift for the human to converge deliberately.
		if _, err := r.git("reset", "--hard", remote); err != nil {
			return err
		}
		r.state.AppliedSHA = remote
		if err := r.saveState(); err != nil {
			return err
		}
		log.Printf("baselined at %.9s without applying (first run)", remote)
	case remote != r.state.AppliedSHA:
		if err := r.apply(r.state.AppliedSHA, remote); err != nil {
			return err
		}
	}

	if err := r.computeDrift(); err != nil {
		return err
	}
	r.publish()
	r.status.Error = ""
	r.status.LastSync = time.Now().Format("2006-01-02 15:04:05")
	return nil
}

// changedRels parses old..new into live-named rel paths (renames become
// a delete plus an add).
func (r *Reconciler) changedRels(old, new_ string) ([]string, error) {
	out, err := r.git("diff", "--name-status", old, new_, "--", r.opts.Subfolder)
	if err != nil {
		return nil, err
	}
	seen := map[string]bool{}
	var rels []string
	prefix := r.opts.Subfolder + "/"
	for _, line := range strings.Split(strings.TrimSpace(out), "\n") {
		if line == "" {
			continue
		}
		fields := strings.Split(line, "\t")
		for _, p := range fields[1:] { // status, path [, path for renames]
			rel := strings.TrimPrefix(p, prefix)
			if rel == secretsSops {
				rel = secretsPlain
			}
			if !seen[rel] {
				seen[rel] = true
				rels = append(rels, rel)
			}
		}
	}
	return rels, nil
}

func (r *Reconciler) apply(old, new_ string) error {
	rels, err := r.changedRels(old, new_)
	if err != nil {
		return err
	}
	var applied []string
	conflicts := map[string]string{}
	backupDir := filepath.Join(r.backupsDir, time.Now().Format("20060102-150405"))

	for _, rel := range rels {
		liveB := r.live(rel)
		was := r.atRef(old, rel)
		if liveB != nil && was != nil && !bytes.Equal(liveB, was) {
			conflicts[rel] = "changed in git and live"
			continue
		}
		want := r.atRef(new_, rel)
		target := filepath.Join(r.configDir, rel)
		if liveB != nil {
			if err := os.MkdirAll(backupDir, 0o700); err != nil {
				return err
			}
			name := strings.ReplaceAll(rel, "/", "__")
			if err := os.WriteFile(filepath.Join(backupDir, name), liveB, 0o600); err != nil {
				return err
			}
		}
		if want == nil { // deleted in git
			_ = os.Remove(target)
		} else {
			if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
				return err
			}
			if err := os.WriteFile(target, want, 0o644); err != nil {
				return err
			}
		}
		applied = append(applied, rel)
	}

	r.status.Conflicts = conflicts
	if len(conflicts) > 0 {
		keys := sortedKeys(conflicts)
		r.ha.Notify("GitOps conflict",
			"Changed in both git and live, not applied: "+strings.Join(keys, ", "))
	}

	if len(applied) > 0 {
		ok, err := r.ha.CoreCheck()
		if err != nil {
			return err
		}
		if !ok {
			for _, rel := range applied { // roll back everything we wrote
				name := strings.ReplaceAll(rel, "/", "__")
				saved, err := os.ReadFile(filepath.Join(backupDir, name))
				target := filepath.Join(r.configDir, rel)
				if err == nil {
					_ = os.WriteFile(target, saved, 0o644)
				} else {
					_ = os.Remove(target)
				}
			}
			r.ha.Notify("GitOps apply rolled back", fmt.Sprintf(
				"%.9s..%.9s failed core config check; live files restored.",
				old, new_))
			r.status.Error = "config check failed; rolled back"
			return nil
		}
		r.activate(applied)
		log.Printf("applied %.9s..%.9s: %v", old, new_, applied)
	}

	r.state.AppliedSHA = new_
	if err := r.saveState(); err != nil {
		return err
	}
	_, err = r.git("reset", "--hard", new_)
	return err
}

// activate reloads what it can and flags a restart for what it can't.
func (r *Reconciler) activate(rels []string) {
	needsRestart := false
	for _, rel := range rels {
		if restartFiles[rel] {
			needsRestart = true
		}
	}
	if needsRestart {
		if r.opts.AutoRestartCore {
			r.ha.CoreRestart()
			r.status.RestartRequired = false
		} else {
			r.status.RestartRequired = true
			r.ha.Notify("GitOps: restart required",
				"configuration.yaml or secrets.yaml changed; restart Home Assistant to activate.")
		}
		return
	}
	reloaded := map[[2]string]bool{}
	allMapped := true
	for _, rel := range rels {
		svc, ok := reloadServices[rel]
		if !ok {
			allMapped = false
			continue
		}
		if !reloaded[svc] {
			r.ha.CallService(svc[0], svc[1], nil)
			reloaded[svc] = true
		}
	}
	if !allMapped {
		r.ha.CallService("homeassistant", "reload_all", nil)
	}
}

func (r *Reconciler) computeDrift() error {
	head := r.state.AppliedSHA
	rels, err := r.desiredRels(head)
	if err != nil {
		return err
	}
	drift := map[string]string{}
	for _, rel := range rels {
		want := r.atRef(head, rel)
		liveB := r.live(rel)
		switch {
		case liveB == nil:
			drift[rel] = "missing live"
		case !bytes.Equal(liveB, want):
			drift[rel] = "modified live"
		}
	}
	r.status.Drift = drift
	return nil
}

func (r *Reconciler) publish() {
	n := len(r.status.Drift) + len(r.status.Conflicts)
	sha := r.state.AppliedSHA
	if len(sha) > 9 {
		sha = sha[:9]
	}
	r.ha.SetState("sensor.ha_gitops_drift", fmt.Sprint(n), map[string]any{
		"friendly_name":    "GitOps drift",
		"icon":             "mdi:source-branch-sync",
		"drift":            sortedKeys(r.status.Drift),
		"conflicts":        sortedKeys(r.status.Conflicts),
		"applied_sha":      sha,
		"restart_required": r.status.RestartRequired,
	})
}

func sortedKeys(m map[string]string) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// ---------- human actions (from the web UI) ----------

func (r *Reconciler) SyncNow() error { return r.Tick() }

func (r *Reconciler) Revert(rel string) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	want := r.atRef(r.state.AppliedSHA, rel)
	target := filepath.Join(r.configDir, rel)
	if want == nil {
		_ = os.Remove(target)
	} else {
		if err := os.WriteFile(target, want, 0o644); err != nil {
			return err
		}
	}
	r.activate([]string{rel})
	if err := r.computeDrift(); err != nil {
		return err
	}
	r.publish()
	return nil
}

func (r *Reconciler) Promote(rels []string, message string) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	for _, rel := range rels {
		liveB := r.live(rel)
		if liveB == nil {
			if _, err := r.git("rm", "-q", r.opts.Subfolder+"/"+rel); err != nil {
				return err
			}
			continue
		}
		if rel == secretsPlain {
			// Re-encrypt with the repo's .sops.yaml recipients so the
			// operator key can still decrypt. Rules match by path.
			out := filepath.Join(r.repoDir, r.opts.Subfolder, secretsSops)
			if err := os.WriteFile(out, liveB, 0o600); err != nil {
				return err
			}
			cmd := exec.Command("sops", "--encrypt", "--in-place", out)
			cmd.Dir = r.repoDir
			cmd.Env = append(os.Environ(), "HOME=/data")
			if o, err := cmd.CombinedOutput(); err != nil {
				return fmt.Errorf("sops encrypt: %v: %s", err, o)
			}
			if _, err := r.git("add", r.opts.Subfolder+"/"+secretsSops); err != nil {
				return err
			}
			continue
		}
		target := filepath.Join(r.repoDir, r.opts.Subfolder, rel)
		if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
			return err
		}
		if err := os.WriteFile(target, liveB, 0o644); err != nil {
			return err
		}
		if _, err := r.git("add", r.opts.Subfolder+"/"+rel); err != nil {
			return err
		}
	}
	if message == "" {
		message = "promote: " + strings.Join(rels, ", ")
	}
	if _, err := r.git("commit", "-m", message); err != nil {
		return err
	}
	if _, err := r.git("pull", "--rebase", "origin", r.opts.Branch); err != nil {
		_, _ = r.git("rebase", "--abort")
		return fmt.Errorf("promote conflicts with new upstream commits; sync first")
	}
	if _, err := r.git("push", "origin", "HEAD:"+r.opts.Branch); err != nil {
		return err
	}
	head, err := r.git("rev-parse", "HEAD")
	if err != nil {
		return err
	}
	r.state.AppliedSHA = strings.TrimSpace(head)
	if err := r.saveState(); err != nil {
		return err
	}
	if err := r.computeDrift(); err != nil {
		return err
	}
	r.publish()
	return nil
}

// DiffText renders a unified diff of git (desired) vs live for the UI.
// Secrets are masked to avoid leaking values into the panel.
func (r *Reconciler) DiffText(rel string) string {
	if rel == secretsPlain {
		return "(secrets diff masked — values changed)"
	}
	r.mu.Lock()
	head := r.state.AppliedSHA
	r.mu.Unlock()

	want := r.atRef(head, rel)
	tmp, err := os.CreateTemp("", "desired-*")
	if err != nil {
		return err.Error()
	}
	defer os.Remove(tmp.Name())
	_, _ = tmp.Write(want)
	_ = tmp.Close()

	livePath := filepath.Join(r.configDir, rel)
	if _, err := os.Stat(livePath); err != nil {
		livePath = os.DevNull
	}
	// git diff --no-index exits 1 when files differ; that's expected.
	cmd := exec.Command("git", "diff", "--no-index",
		"--src-prefix=git/", "--dst-prefix=live/", tmp.Name(), livePath)
	out, _ := cmd.Output()
	return string(out)
}

File: ha-git-ops/rootfs/opt/ha_gitops/reconciler.py
"""Reconciler: git desired state vs live /config, ArgoCD-style.

State machine per managed file:
  in-sync    live == HEAD
  drift      live != HEAD, and HEAD did not move (click-ops change)
  conflict   HEAD moved AND live != the previously-applied version
             (both sides changed; never auto-resolved)

Apply only writes files whose live content still matches the previously
applied git version — local edits are never stomped. On first run nothing
is written at all: the remote HEAD becomes the baseline and every
difference shows up as drift for the human to promote or revert.
"""

import json
import logging
import shutil
import subprocess
import time
from pathlib import Path

log = logging.getLogger("reconciler")

SECRETS_SOPS = "secrets.sops.yaml"
SECRETS_PLAIN = "secrets.yaml"
# Files whose change requires a core restart rather than a reload.
RESTART_FILES = {"configuration.yaml", SECRETS_PLAIN}
RELOAD_SERVICES = {
    "automations.yaml": ("automation", "reload"),
    "scenes.yaml": ("scene", "reload"),
    "scripts.yaml": ("script", "reload"),
}


class Reconciler:
    def __init__(self, opts, ha):
        self.opts = opts
        self.ha = ha
        self.repo = Path("/data/repo")
        self.config = Path("/homeassistant")
        self.state_file = Path("/data/state.json")
        self.age_key = Path("/data/age.key")
        self.ssh_key = Path("/data/ssh/id_ed25519")
        self.backups = Path("/data/backups")
        self.state = self._load_state()
        # Computed each tick; read by the web UI.
        self.status = {"drift": {}, "conflicts": {}, "error": None,
                       "restart_required": False, "last_sync": None}

    # ---------- setup ----------

    def _load_state(self):
        if self.state_file.exists():
            return json.loads(self.state_file.read_text())
        return {"applied_sha": None}

    def _save_state(self):
        self.state_file.write_text(json.dumps(self.state))

    def ensure_keys(self):
        if not self.age_key.exists():
            subprocess.run(["age-keygen", "-o", str(self.age_key)], check=True)
            self.age_key.chmod(0o600)
        age_pub = subprocess.run(
            ["age-keygen", "-y", str(self.age_key)],
            check=True, capture_output=True, text=True).stdout.strip()
        if not self.ssh_key.exists():
            self.ssh_key.parent.mkdir(parents=True, exist_ok=True)
            subprocess.run(
                ["ssh-keygen", "-t", "ed25519", "-N", "", "-C", "ha-git-ops",
                 "-f", str(self.ssh_key)], check=True)
        ssh_pub = (self.ssh_key.with_suffix(".pub")).read_text().strip()
        self.pubkeys = {"age": age_pub, "ssh": ssh_pub}
        log.info("age public key (add to .sops.yaml recipients, then "
                 "`sops updatekeys`): %s", age_pub)
        log.info("ssh deploy key (add to the git host with write access): %s",
                 ssh_pub)

    # ---------- git plumbing ----------

    def _git(self, *args, check=True, input=None):
        env = {
            "GIT_SSH_COMMAND":
                f"ssh -i {self.ssh_key} -o StrictHostKeyChecking=accept-new",
            "HOME": "/data",
            "PATH": "/usr/local/bin:/usr/bin:/bin",
        }
        r = subprocess.run(["git", "-C", str(self.repo), *args],
                           capture_output=True, text=True, env=env, input=input)
        if check and r.returncode != 0:
            raise RuntimeError(f"git {' '.join(args)}: {r.stderr.strip()}")
        return r.stdout

    def ensure_repo(self):
        if (self.repo / ".git").exists():
            return
        if not self.opts["repo_url"]:
            raise RuntimeError("repo_url is not configured")
        self.repo.parent.mkdir(parents=True, exist_ok=True)
        env = {
            "GIT_SSH_COMMAND":
                f"ssh -i {self.ssh_key} -o StrictHostKeyChecking=accept-new",
            "HOME": "/data",
            "PATH": "/usr/local/bin:/usr/bin:/bin",
        }
        r = subprocess.run(
            ["git", "clone", "--filter=blob:none", "--no-checkout",
             "--branch", self.opts["branch"],
             self.opts["repo_url"], str(self.repo)],
            capture_output=True, text=True, env=env)
        if r.returncode != 0:
            raise RuntimeError(f"git clone: {r.stderr.strip()}")
        self._git("sparse-checkout", "set", self.opts["subfolder"])
        self._git("checkout", self.opts["branch"])
        self._git("config", "user.name", self.opts["commit_name"])
        self._git("config", "user.email", self.opts["commit_email"])

    def _tracked(self, ref):
        """Managed rel-paths (relative to subfolder) at a git ref."""
        sub = self.opts["subfolder"]
        out = self._git("ls-tree", "-r", "--name-only", ref, "--", sub)
        return [p[len(sub) + 1:] for p in out.splitlines() if p.strip()]

    def _at_ref(self, ref, rel):
        """File content at a git ref, or None. Decrypts the secrets file."""
        sub = self.opts["subfolder"]
        src = SECRETS_SOPS if rel == SECRETS_PLAIN else rel
        r = subprocess.run(
            ["git", "-C", str(self.repo), "show", f"{ref}:{sub}/{src}"],
            capture_output=True)
        if r.returncode != 0:
            return None
        if rel == SECRETS_PLAIN:
            return self._decrypt(r.stdout)
        return r.stdout

    def _decrypt(self, blob):
        r = subprocess.run(
            ["sops", "--decrypt", "--input-type", "yaml",
             "--output-type", "yaml", "/dev/stdin"],
            input=blob, capture_output=True,
            env={"SOPS_AGE_KEY_FILE": str(self.age_key),
                 "PATH": "/usr/local/bin:/usr/bin:/bin"})
        if r.returncode != 0:
            raise RuntimeError(f"sops decrypt failed: {r.stderr.decode()}")
        return r.stdout

    def desired_rels(self, ref):
        """Managed rel-paths as they appear live (sops name translated)."""
        rels = []
        for rel in self._tracked(ref):
            rels.append(SECRETS_PLAIN if rel == SECRETS_SOPS else rel)
        return rels

    def _live(self, rel):
        p = self.config / rel
        return p.read_bytes() if p.exists() else None

    # ---------- reconcile ----------

    async def tick(self):
        self.ensure_repo()
        self._git("fetch", "origin", self.opts["branch"])
        remote = self._git("rev-parse",
                           f"origin/{self.opts['branch']}").strip()
        applied = self.state["applied_sha"]

        if applied is None:
            # First run: baseline only, write nothing. Differences appear
            # as drift for the human to converge deliberately.
            self._git("reset", "--hard", remote)
            self.state["applied_sha"] = remote
            self._save_state()
            log.info("baselined at %s without applying (first run)", remote[:9])
        elif remote != applied:
            await self._apply(applied, remote)

        self._compute_drift()
        await self._publish()
        self.status["last_sync"] = time.strftime("%Y-%m-%d %H:%M:%S")

    async def _apply(self, old, new):
        """Apply old..new to /config, skipping conflicted files."""
        changed = [l.split("\t", 1) for l in self._git(
            "diff", "--name-status", old, new, "--",
            self.opts["subfolder"]).splitlines()]
        sub = self.opts["subfolder"]
        applied, conflicts = [], {}
        backup_dir = self.backups / time.strftime("%Y%m%d-%H%M%S")

        for status_, path in changed:
            rel = path[len(sub) + 1:]
            if rel == SECRETS_SOPS:
                rel = SECRETS_PLAIN
            live = self._live(rel)
            was = self._at_ref(old, rel)
            if live is not None and was is not None and live != was:
                conflicts[rel] = "changed in git and live"
                continue
            want = self._at_ref(new, rel)
            target = self.config / rel
            if live is not None:
                backup_dir.mkdir(parents=True, exist_ok=True)
                (backup_dir / rel.replace("/", "__")).write_bytes(live)
            if want is None:  # deleted in git
                if target.exists():
                    target.unlink()
            else:
                target.parent.mkdir(parents=True, exist_ok=True)
                target.write_bytes(want)
            applied.append(rel)

        self.status["conflicts"] = conflicts
        if conflicts:
            await self.ha.notify(
                "GitOps conflict",
                "Changed in both git and live, not applied: "
                + ", ".join(sorted(conflicts)))

        if applied:
            ok = await self.ha.core_check()
            if not ok:
                for rel in applied:  # roll back everything we wrote
                    saved = backup_dir / rel.replace("/", "__")
                    target = self.config / rel
                    if saved.exists():
                        target.write_bytes(saved.read_bytes())
                    elif target.exists():
                        target.unlink()
                await self.ha.notify(
                    "GitOps apply rolled back",
                    f"{old[:9]}..{new[:9]} failed core config check; "
                    "live files restored.")
                self.status["error"] = "config check failed; rolled back"
                return
            await self._activate(applied)
            log.info("applied %s..%s: %s", old[:9], new[:9], applied)

        self.state["applied_sha"] = new
        self._save_state()
        self._git("reset", "--hard", new)

    async def _activate(self, rels):
        """Reload what we can; flag a restart for what we can't."""
        if any(r in RESTART_FILES for r in rels):
            if self.opts["auto_restart_core"]:
                await self.ha.core_restart()
                self.status["restart_required"] = False
            else:
                self.status["restart_required"] = True
                await self.ha.notify(
                    "GitOps: restart required",
                    "configuration.yaml or secrets.yaml changed; restart "
                    "Home Assistant to activate.")
            return
        reloaded = set()
        for rel in rels:
            svc = RELOAD_SERVICES.get(rel)
            if svc and svc not in reloaded:
                await self.ha.call_service(*svc)
                reloaded.add(svc)
        if not all(RELOAD_SERVICES.get(r) for r in rels):
            await self.ha.call_service("homeassistant", "reload_all")

    def _compute_drift(self):
        head = self.state["applied_sha"]
        drift = {}
        for rel in self.desired_rels(head):
            want = self._at_ref(head, rel)
            live = self._live(rel)
            if live is None:
                drift[rel] = "missing live"
            elif live != want:
                drift[rel] = "modified live"
        self.status["drift"] = drift

    async def _publish(self):
        n = len(self.status["drift"]) + len(self.status["conflicts"])
        await self.ha.set_state(
            "sensor.ha_gitops_drift", n,
            {"friendly_name": "GitOps drift", "icon": "mdi:source-branch-sync",
             "drift": sorted(self.status["drift"]),
             "conflicts": sorted(self.status["conflicts"]),
             "applied_sha": (self.state["applied_sha"] or "")[:9],
             "restart_required": self.status["restart_required"]})

    # ---------- human actions (from the web UI) ----------

    async def revert(self, rel):
        want = self._at_ref(self.state["applied_sha"], rel)
        target = self.config / rel
        if want is None:
            if target.exists():
                target.unlink()
        else:
            target.write_bytes(want)
        await self._activate([rel])
        self._compute_drift()
        await self._publish()

    async def promote(self, rels, message):
        sub = self.opts["subfolder"]
        for rel in rels:
            live = self._live(rel)
            if live is None:
                self._git("rm", "-q", f"{sub}/{rel}")
                continue
            if rel == SECRETS_PLAIN:
                # Re-encrypt with the repo's .sops.yaml recipients so the
                # operator key can still decrypt. Rules match by path.
                out = self.repo / sub / SECRETS_SOPS
                out.write_bytes(live)
                r = subprocess.run(
                    ["sops", "--encrypt", "--in-place", str(out)],
                    capture_output=True, cwd=str(self.repo),
                    env={"PATH": "/usr/local/bin:/usr/bin:/bin"})
                if r.returncode != 0:
                    raise RuntimeError(f"sops encrypt: {r.stderr.decode()}")
                self._git("add", f"{sub}/{SECRETS_SOPS}")
                continue
            target = self.repo / sub / rel
            target.parent.mkdir(parents=True, exist_ok=True)
            target.write_bytes(live)
            self._git("add", f"{sub}/{rel}")
        self._git("commit", "-m", message or f"promote: {', '.join(rels)}")
        try:
            self._git("pull", "--rebase", "origin", self.opts["branch"])
        except RuntimeError:
            self._git("rebase", "--abort", check=False)
            raise RuntimeError(
                "promote conflicts with new upstream commits; sync first")
        self._git("push", "origin", f"HEAD:{self.opts['branch']}")
        self.state["applied_sha"] = self._git("rev-parse", "HEAD").strip()
        self._save_state()
        self._compute_drift()
        await self._publish()
